File: python/problems/LeetCode_1_Two_Sum.py
## brute force method
## run two loops to check indexes
## T.C -> O(n^2)
## S.C O(1)

def brute_force( nums:list, target: int):
        small=0
        large=0
        for i in range(0,len(nums)):
            for j in range(i+1,len(nums)):
                ## i+1 is reason this brute force method a bit efficient we dont check backawards
                if(i==j):
                    continue
                elif(nums[i]+nums[j]==target):
                        small=i
                        large=j
                        break
                        # small=min(i,j)
                        # large=max(i,j)
                        ## i should be the smaller by default
        return [small,large]

## lets try an hashmap method
## we simply do like for example 8 + _ =14 so we just find if the 14-8 elements is present in hashmap(diict in python)
## T.C -> O(n)(in c++ maps -> O(nlogn) & unordered_sets -> O(n))
## S.C -> O(n)
def better_approach(nums:list,target:int):
    my_dict=dict()
    for i in range(len(nums)):
          complement=target-nums[i]
          if complement in my_dict:
                return [my_dict[complement],i] # indexes
          my_dict[nums[i]] = i
          

    
nums=[2,7,11,15]
print(brute_force(nums,9))
print(better_approach(nums,9))


## note: there is an two sum II problem on leetcode
# it is already sorted so can we directly use two pointers approach on it 

File: cpp/problems/LeetCode_1_Two_Sum.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
using namespace std;

// question description
// Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

// brute force
// ## brute force method
// run two loops to check indexes
// T.C -> O(n^2)
// S.C O(1)

vector<int> brute_method(const vector<int> &nums, int t)
{
    vector<int> idx;
    for (int i = 0; i < nums.size(); ++i)
    {
        for (int j = i; j < nums.size(); ++j)
        {
            if (nums[i] + nums[j] == t)
            {
                idx.push_back(i);
                idx.push_back(j);
            }
        }
    }
    return idx;
}

// lets try an hashmap method
// we simply do like for example 8 + _ =14 so we just find if the 14-8 elements is present in hashmap(diict in python)
// T.C -> O(n)(in c++ maps -> O(nlogn) & unordered_sets -> O(n))
// S.C -> O(n)


vector<int> optimal_method(const vector<int> &nums, int t)
{
    vector<int> idx;
    unordered_map<int,int> mapp;
    for(int i=0;i<nums.size();i++){
        mapp[nums[i]]=i;

    }
    for(int i=0;i<nums.size();i++){
        int c=t-nums[i];
        if(mapp.find(c)!=mapp.end()){
            idx.push_back(i);
            idx.push_back(mapp[c]);
            break;
        }

    }
    return idx;
}

int main()
{
    vector<int> nums = {11, 7, 2, 15};
    int target = 9;
    for (auto item : brute_method(nums, target))
    {
        cout << item << " ";
    }
    cout << "\n";
    for (auto item : optimal_method(nums, target))
    {
        cout << item << " ";
    }

    return 0;
}